File: Tests/MCPEnablerCoreTests/FormMapperTests.swift
import XCTest
@testable import MCPEnablerCore

final class FormMapperTests: XCTestCase {
    func testCleanLocalConfigIsLossless() {
        let analysis = FormMapper.analyze(.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("pkg")]),
            "env": .object(["KEY": .string("v")]),
        ]))
        XCTAssertTrue(analysis.isLossless)
        XCTAssertEqual(analysis.model,
                       FormModel(command: "npx", args: ["-y", "pkg"],
                                 env: ["KEY": "v"], additional: [:]))
    }

    func testUnknownKeysArePreservedNotLost() {
        let analysis = FormMapper.analyze(.object([
            "command": .string("x"),
            "type": .string("http"),
            "headers": .object(["Authorization": .string("Bearer t")]),
        ]))
        XCTAssertTrue(analysis.isLossless)
        XCTAssertEqual(Set(analysis.model.additional.keys), ["type", "headers"])
    }

    func testStructuralViolationsAreListedAsLost() {
        let analysis = FormMapper.analyze(.object([
            "command": .int(5),
            "args": .array([.string("ok"), .object([:]), .int(3)]),
            "env": .object(["GOOD": .string("y"), "BAD": .int(1)]),
        ]))
        XCTAssertFalse(analysis.isLossless)
        XCTAssertEqual(analysis.lost, ["args[1] (object)", "args[2] (number)",
                                       "command (number)", "env.BAD (number)"])
        XCTAssertEqual(analysis.model.args, ["ok"])
        XCTAssertEqual(analysis.model.env, ["GOOD": "y"])
        XCTAssertEqual(analysis.model.command, "")
    }

    func testNonObjectArgsOrEnvAreLost() {
        let analysis = FormMapper.analyze(.object([
            "command": .string("x"),
            "args": .string("not an array"),
            "env": .array([]),
        ]))
        XCTAssertEqual(analysis.lost, ["args (not an array)", "env (not an object)"])
    }

    func testNonObjectConfigIsEntirelyLost() {
        let analysis = FormMapper.analyze(.string("weird"))
        XCTAssertEqual(analysis.lost, ["entire configuration (not a JSON object)"])
    }

    func testSerializeRoundTripsLosslessConfig() {
        let original = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("pkg")]),
            "env": .object(["K": .string("v")]),
            "headers": .object(["H": .string("x")]),
        ])
        let analysis = FormMapper.analyze(original)
        XCTAssertEqual(FormMapper.serialize(analysis.model), original)
    }

    func testSerializeOmitsEmptyArgsAndEnv() {
        let value = FormMapper.serialize(
            FormModel(command: "swift", args: [], env: [:], additional: [:]))
        XCTAssertEqual(value, .object(["command": .string("swift")]))
    }

    func testCommandlessConfigRoundTripsWithoutInjectingCommand() {
        let original = JSONValue.object([
            "type": .string("http"),
            "url": .string("https://example.com/mcp"),
        ])
        let analysis = FormMapper.analyze(original)
        XCTAssertTrue(analysis.isLossless)
        XCTAssertEqual(FormMapper.serialize(analysis.model), original)
    }
}

File: windows/tests/ConnectorControl.Core.Tests/FormMapperTests.cs
namespace ConnectorControl.Core.Tests;

public class FormMapperTests
{
    private static Dictionary<string, string> Env(params (string, string)[] pairs) =>
        pairs.ToDictionary(p => p.Item1, p => p.Item2, StringComparer.Ordinal);

    [Fact]
    public void CleanLocalConfigIsLossless()
    {
        var analysis = FormMapper.Analyze(JsonValue.Object(
            ("command", JsonValue.String("npx")),
            ("args", JsonValue.Array([JsonValue.String("-y"), JsonValue.String("pkg")])),
            ("env", JsonValue.Object(("KEY", JsonValue.String("v"))))));
        Assert.True(analysis.IsLossless);
        Assert.Equal(new FormModel("npx", ["-y", "pkg"], Env(("KEY", "v"))), analysis.Model);
    }

    [Fact]
    public void UnknownKeysArePreservedNotLost()
    {
        var analysis = FormMapper.Analyze(JsonValue.Object(
            ("command", JsonValue.String("x")),
            ("type", JsonValue.String("http")),
            ("headers", JsonValue.Object(("Authorization", JsonValue.String("Bearer t"))))));
        Assert.True(analysis.IsLossless);
        Assert.Equal(new HashSet<string> { "type", "headers" }, analysis.Model.Additional.Keys.ToHashSet());
    }

    [Fact]
    public void StructuralViolationsAreListedAsLost()
    {
        var analysis = FormMapper.Analyze(JsonValue.Object(
            ("command", JsonValue.Int(5)),
            ("args", JsonValue.Array([JsonValue.String("ok"), JsonValue.Object(), JsonValue.Int(3)])),
            ("env", JsonValue.Object(("GOOD", JsonValue.String("y")), ("BAD", JsonValue.Int(1))))));
        Assert.False(analysis.IsLossless);
        Assert.Equal(["args[1] (object)", "args[2] (number)", "command (number)", "env.BAD (number)"], analysis.Lost);
        Assert.Equal(["ok"], analysis.Model.Args);
        Assert.Equal(Env(("GOOD", "y")), analysis.Model.Env);
        Assert.Equal("", analysis.Model.Command);
    }

    [Fact]
    public void NonObjectArgsOrEnvAreLost()
    {
        var analysis = FormMapper.Analyze(JsonValue.Object(
            ("command", JsonValue.String("x")),
            ("args", JsonValue.String("not an array")),
            ("env", JsonValue.Array([]))));
        Assert.Equal(["args (not an array)", "env (not an object)"], analysis.Lost);
    }

    [Fact]
    public void NonObjectConfigIsEntirelyLost()
    {
        Assert.Equal(["entire configuration (not a JSON object)"], FormMapper.Analyze(JsonValue.String("weird")).Lost);
    }

    [Fact]
    public void SerializeRoundTripsLosslessConfig()
    {
        var original = JsonValue.Object(
            ("command", JsonValue.String("npx")),
            ("args", JsonValue.Array([JsonValue.String("-y"), JsonValue.String("pkg")])),
            ("env", JsonValue.Object(("K", JsonValue.String("v")))),
            ("headers", JsonValue.Object(("H", JsonValue.String("x")))));
        Assert.Equal(original, FormMapper.Serialize(FormMapper.Analyze(original).Model));
    }

    [Fact]
    public void SerializeOmitsEmptyArgsAndEnv()
    {
        Assert.Equal(JsonValue.Object(("command", JsonValue.String("swift"))), FormMapper.Serialize(new FormModel("swift")));
    }

    [Fact]
    public void CommandlessConfigRoundTripsWithoutInjectingCommand()
    {
        var original = JsonValue.Object(("type", JsonValue.String("http")), ("url", JsonValue.String("https://example.com/mcp")));
        var analysis = FormMapper.Analyze(original);
        Assert.True(analysis.IsLossless);
        Assert.Equal(original, FormMapper.Serialize(analysis.Model));
    }
}
